### ISSUE TITLE
[JENKINS-3966] Add option to clean workspace before each build

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/jenkins#18889]
<p>If this option is selected the workspace will be deleted before each build<br/>
begins so that each build can have a fresh workspace.  This will be off by default.</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kaxelson">kaxelson</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-3966" target="_blank">Add option to clean workspace before each build</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kaxelson">kaxelson</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: core
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2014-05-23T15:20:46+00:00
<li><b>votes</b>: 19
<li><b>watchers</b>: 21
<li><b>imported</b>: 2025-11-24
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>If this option is selected the workspace will be deleted before each build
begins so that each build can have a fresh workspace.  This will be off by default.</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Platform: All, OS: All</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kohsuke.png" title="kohsuke's avatar" /> kohsuke:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-3966?focusedId=131371&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-131371">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Rolled back from 1.315 release in rev.19544. See
<a href="http://www.nabble.com/attention-all-subversion-users-td24335693.html" class="external-link" target="_blank" rel="nofollow noopener">http://www.nabble.com/attention-all-subversion-users-td24335693.html</a> for the
discussion.</p>
</pre>
</details>

</ul>
</details>

<p>Rolled back from 1.315 release in rev.19544. See<br/>
<a href="http://www.nabble.com/attention-all-subversion-users-td24335693.html" class="external-link" target="_blank" rel="nofollow noopener">http://www.nabble.com/attention-all-subversion-users-td24335693.html</a> for the<br/>
discussion.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-3966?focusedId=131372&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-131372">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Note that the Mercurial plugin already has an option to delete all nonversioned
files from the workspace before starting. (The .hg/ subdirectory is excluded as
is usual.) I think there is a similar issue open for Subversion. This is more
efficient than deleting everything in the workspace, but may not be applicable
for all situations.</p>
</pre>
</details>

</ul>
</details>

<p>Note that the Mercurial plugin already has an option to delete all nonversioned<br/>
files from the workspace before starting. (The .hg/ subdirectory is excluded as<br/>
is usual.) I think there is a similar issue open for Subversion. This is more<br/>
efficient than deleting everything in the workspace, but may not be applicable<br/>
for all situations.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>dma_k:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-3966?focusedId=131373&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-131373">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The option to remove all workspace is important to make maven release plugin to
work correctly for the projects, that generate sources e.g. with JAXB. In my
case I commit JAXB sources from Windows environment, which after being
re-generated have different EOF and SVN senses this as uncommitted change, which
in its turn blocks the release.
If the fix is ready, what is the target milestone for this fix?</p>
</pre>
</details>

</ul>
</details>

<p>The option to remove all workspace is important to make maven release plugin to<br/>
work correctly for the projects, that generate sources e.g. with JAXB. In my<br/>
case I commit JAXB sources from Windows environment, which after being<br/>
re-generated have different EOF and SVN senses this as uncommitted change, which<br/>
in its turn blocks the release.<br/>
If the fix is ready, what is the target milestone for this fix?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>smr99:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-3966?focusedId=134600&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-134600">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>What is the status of this?  I read through the referenced email thread (now at <a href="http://markmail.org/message/mcgftdni6cvpca7v" class="external-link" target="_blank" rel="nofollow noopener">http://markmail.org/message/mcgftdni6cvpca7v</a>) which seemed to end with agreement on the feature but issues with backwards compatibility.</p>

</pre>
</details>

</ul>
</details>

<p>What is the status of this?  I read through the referenced email thread (now at <a href="http://markmail.org/message/mcgftdni6cvpca7v" class="external-link" target="_blank" rel="nofollow noopener">http://markmail.org/message/mcgftdni6cvpca7v</a>) which seemed to end with agreement on the feature but issues with backwards compatibility.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jhoomsharabi:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-3966?focusedId=135227&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-135227">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Does this functionality anytime implemented, as I see this issue is being reopened. </p>
</pre>
</details>

</ul>
</details>

<p>Does this functionality anytime implemented, as I see this issue is being reopened. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/rpetti.png" title="rpetti's avatar" /> rpetti:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-3966?focusedId=136487&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-136487">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Can't this option just call the same function as the "Wipe Out Workspace" feature? If the SCM plugin supports it, then everything will be fine. If it doesn't then it simply won't delete the workspace.</p>
</pre>
</details>

</ul>
</details>

<p>Can't this option just call the same function as the "Wipe Out Workspace" feature? If the SCM plugin supports it, then everything will be fine. If it doesn't then it simply won't delete the workspace.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-3966?focusedId=148065&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-148065">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Kohsuke Kawaguchi
Path:
 core/src/main/java/hudson/scm/NoCleaner.java
 core/src/main/java/hudson/scm/RmRfCleaner.java
 core/src/main/java/hudson/scm/WorkspaceCleaner.java
 core/src/main/java/hudson/scm/WorkspaceCleanerDescriptor.java
 core/src/main/resources/hudson/scm/SCM/configure-workspaceCleaner.jellytag
 core/src/main/resources/hudson/scm/WorkspaceCleaner/config.jelly
<a href="http://jenkins-ci.org/commit/jenkins/07c6c4f1dc9c41a2cf8fc1ee9a8935310ca9afb0" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/07c6c4f1dc9c41a2cf8fc1ee9a8935310ca9afb0</a>
Log:
<a href="https://issues.jenkins.io/browse/JENKINS-3966" title="Add option to clean workspace before each build" class="issue-link" data-issue-key="JENKINS-3966"><del>JENKINS-3966</del></a> added a new extension point for workspace cleaner</p>



</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Kohsuke Kawaguchi<br/>
Path:<br/>
 core/src/main/java/hudson/scm/NoCleaner.java<br/>
 core/src/main/java/hudson/scm/RmRfCleaner.java<br/>
 core/src/main/java/hudson/scm/WorkspaceCleaner.java<br/>
 core/src/main/java/hudson/scm/WorkspaceCleanerDescriptor.java<br/>
 core/src/main/resources/hudson/scm/SCM/configure-workspaceCleaner.jellytag<br/>
 core/src/main/resources/hudson/scm/WorkspaceCleaner/config.jelly<br/>
<a href="http://jenkins-ci.org/commit/jenkins/07c6c4f1dc9c41a2cf8fc1ee9a8935310ca9afb0" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/07c6c4f1dc9c41a2cf8fc1ee9a8935310ca9afb0</a><br/>
Log:<br/>
<a href="https://issues.jenkins.io/browse/JENKINS-3966" title="Add option to clean workspace before each build" class="issue-link" data-issue-key="JENKINS-3966"><del>JENKINS-3966</del></a> added a new extension point for workspace cleaner</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jlpinardon.png" title="jlpinardon's avatar" /> jlpinardon:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-3966?focusedId=148074&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-148074">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>It would be a user friendly enhancement to add the ability to select a list (possibly based on pattern) of directory to  wipe out (or not).</p>
</pre>
</details>

</ul>
</details>

<p>It would be a user friendly enhancement to add the ability to select a list (possibly based on pattern) of directory to  wipe out (or not).</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>johnstosh:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-3966?focusedId=149936&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-149936">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>(I've been wiping out specific directories by adding ant tasks into the my jobs.)</p>

<p>Hey guys, thanks for enhancing this section of Jenkins.  The clean selections in Jenkins 1.414 work great for me: both the delete-workspace and the other varieties of cleaning.</p>
</pre>
</details>

</ul>
</details>

<p>(I've been wiping out specific directories by adding ant tasks into the my jobs.)</p>

<p>Hey guys, thanks for enhancing this section of Jenkins.  The clean selections in Jenkins 1.414 work great for me: both the delete-workspace and the other varieties of cleaning.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-3966?focusedId=201783&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-201783">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The mentioned plugin should be used; or many SCM plugins now offer the option of deleting unversioned files.</p>
</pre>
</details>

</ul>
</details>

<p>The mentioned plugin should be used; or many SCM plugins now offer the option of deleting unversioned files.</p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `relates_to` from Jira: <a href="https://github.com/jenkinsci/jenkins/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-3934%22">JENKINS-3934</a>]</i>
